client/image_list_test.go

```diff
@@ -158,3 +159,66 @@ func TestImageListApiBefore125(t *testing.T) {
 		t.Fatalf("expected 2 images, got %v", images)
 	}
 }
+
+// Checks if shared-size query parameter is set/not being set correctly
+// for /images/json.
+func TestImageListWithSharedSize(t *testing.T) {
+	// performs ImageList call and returns the sent url query
+	getImageListAndGetQuery := func(version string, options types.ImageListOptions) (query url.Values, err error) {
+		client := &Client{
+			client: newMockClient(func(req *http.Request) (*http.Response, error) {
+				query = req.URL.Query()
+
+				return &http.Response{
+					StatusCode: http.StatusOK,
+					Body:       io.NopCloser(strings.NewReader("[]")),
+				}, nil
+			}),
+			version: version,
+		}
+
+		_, err = client.ImageList(context.Background(), options)
+		return query, err
+	}
+
+	const sharedSize = "shared-size"
+
+	// Check if the shared-size query parameter is unset
+	for _, tC := range []struct {
+		name    string
+		version string
+		options types.ImageListOptions
+	}{
+		{name: "after 1.42, if not requested", version: "1.42", options: types.ImageListOptions{SharedSize: false}},
+		{name: "before 1.42, even if requested", version: "1.41", options: types.ImageListOptions{SharedSize: true}},
+	} {
+		t.Run("unset "+tC.name, func(t *testing.T) {
+			query, err := getImageListAndGetQuery(tC.version, tC.options)
+			if err != nil {
+				t.Error(err)
+			}
+
+			if query.Has(sharedSize) {
+				t.Errorf("expected %q to be unset", sharedSize)
+			}
+		})
+	}
+
+	t.Run("set after 1.42, if requested", func(t *testing.T) {
+		query, err := getImageListAndGetQuery("1.42", types.ImageListOptions{SharedSize: true})
```

Comment:
It looks like this test is mostly identical to the one inside the table above, with the exception of checking the value that's set in the query-string.

If we add a (e.g.) `sharedSize string` to the table, we can use that string to;

- if not-empty: check if the query-arg is set (and vice-versa)
- check the value of the query-param (we can do that unconditionally, as `query.Get()` returns an empty string if the value is set (but empty) or not set.

That way we can have all tests in the table, and we can also get rid of the `getImageListAndGetQuery` and inline the code.

Reply:
I kept them separate to avoid different behaviour check depending on the table value (`Has` vs `!Has`). But looking at your PR, I think it's indeed a bit more readable, so let's go with this one, thanks!